### ISSUE TITLE
Why individual images instead of sprite maps

### ISSUE THREAD

**Issue by jakemac53** [adityaravishankar/command-and-conquer#26]
Instead of having an individual image for every single frame of animation for every object, you should make just one image for each object that holds all the frames. Then you just change the part of the image that you actually display for each frame. This would help your load time issues since it would dramatically reduce http requests.

**Comment by jakemac53** [comment]
this can be done using the same drawImage function as you currently use, but instead of just supplying the image and an x/y following this format:

context.drawImage(image, dx, dy)

You can supply more arguments to tell it what portion of the image to draw which looks like this:

context.drawImage(image, sx, sy, sw, sh, dx, dy, dw, dh)

For more info visite here http://www.w3.org/TR/2dcontext/#images.

**Comment by jakemac53** [comment]
I have forked the repo and I might take a shot at implementing this in your game so we can see how much it helps.

**Comment by jakemac53** [comment]
Ok I have implemented this for all the vehicles, you can see it in action at http://conquer.jacobmacdonald.com. You won't actually notice a difference when playing it, but there are already a ton less http requests (160 for vehicles down to 11)... is this something you would like me to implement for the other objects? If so I am happy to and submit a pull request once I have completed it... otherwise I won't waste my time ;).

**Comment by adityaravishankar** [comment]
Yes. This is a high priority.

**Comment by adityaravishankar** [comment]
@jakemac53 I've also been advised to use PNG instead of GIF for the sprite sheets since the image size is significantly smaller.... Could you use PNGs instead?

**Comment by deanc** [comment]
It depends on the image, but in this case it's most likely PNG will produce the smallest file size.

Jacob, your link doesn't work.

**Comment by adityaravishankar** [comment]
I created the image maps in both gif and png... PNGs are as littls as 1/5 th of the GIF sizes in some cases

**Comment by adityaravishankar** [comment]
@deanc the url works fine for me.

**Comment by jakemac53** [comment]
I will proceed with this and change to use png's then? 

@deanc the link was probably just still loading i stripped down the index.html file a lot so it would load faster and broke the loading screen (I think it was like a css bg or something?). In any case until it loads you just get a blank screen on there. I will revert all my changes to the file before I submit a pull request.

**Comment by jakemac53** [comment]
I just tested gif vs png-8 for the mcv and size difference seems to be pretty negligible (14382 bytes(png) compared to 14512 bytes(gif)). You can see the 2 images for the mcv here http://conquer.jacobmacdonald.com/images/units/vehicles/mcv/.  I am not super familiar with image compression though I just used the default png-8 settings under save for web in photoshop. For now I will just use these same settings for png-8 since it is slightly smaller than the gif but if somebody more knowledgable on image compression had some suggestions it would be nice to hear those.

Also I have only actually done this for the mcv (I will test more once im off work). Maybe I will get better results with the others.

**Comment by adityaravishankar** [comment]
Thanks... The compression shows up more when you are doing building sprites
which are a little larger than the unit sprites.....

**Comment by deanc** [comment]
Is there any reason why all vehicle animations couldn't be on the same sprite? MCV's, tanks, harvesters etc.?

**Comment by jakemac53** [comment]
The entire sprite image gets loaded and stored in memory for each type of object. If all objects were on the same sprite this would be a lot of memory overhead especially as there were more and more objects. Also, if you do it that way its much harder to maintain (if you want to change one object by say adding a frame, it would change the position of all the other sprites and they would all have to be updated later).

**Comment by jakemac53** [comment]
I guess you could just load one sprite into memory though and then they could all just have a reference to it..... but still from a maintainability standpoint its easier to do them separate and the # of files is still greatly reduced.

**Comment by adityaravishankar** [comment]
Well keeping them in separate files just makes it easier to add/remove
units....

Otherwise we will need to remember which order we stored the units in, and
calculate a separate offset depending on which unit we want to display....
The fact that each unit has a different height/width means it becomes more
work calculating the correct offset..... ..

It can be done though...If someone has a clean solution for that, I don't
see why not....If not, is it worth the effort?

**Comment by jakemac53** [comment]
What if we did do one sprite and each type object had its own row (tiled horizontally). Then we would only need to store a y-offset for each type of object and adding frames to a single sprite would not affect the sprite positioning of other objects? The only issue with this that I see is it would result in lots of empty space in the sprite (it would be as wide as the object with the most # of frames*frame width). I wonder if this empty space really matters though?

**Comment by adityaravishankar** [comment]
The way I designed the levels, each level decides what units and buildings
the player has access to for the level... So if I only have infantry for
level 1, I won't load tanks and plane sprites....

Also, this way, we can do creative lazy loading... For example, load tanks
only while factory is being built so that tanks get loaded by the time the
factory is built..... Or load factory only when the first refinery is
built.....etc....

**Comment by jakemac53** [comment]
Also, while having a ton of images is not a good idea, having a few dozen isn't necessarily bad because it allows the browser to download them in parallel as opposed to waiting on one large image.

**Comment by jakemac53** [comment]
I think I will wait until tomorrow to actually implement this to see if there is some more feedback that sways the decision? I think currently we are leaning towards the one image per unit type?

Also, currently there are multiple statuses for each unit with their own sprites. How I have currently implemented it is that each of these statuses also has its own sprite because this is the easiest way to implement it within the current system. However I could combine these sprites so there is exactly one sprite for each unit type. For the vehicles the only one that really has more than 2 "statuses" is the harvester, which has 8 different harvesting statuses (which actually represent the 8 directions it can face). Because of this it actually has 9 total sprites where the light-tank has 2 and the mcv has 1.

I think at the very least I will combine these "status" sprites all into one sprite for each unit, with each "status" on their own line in the sprite

**Comment by adityaravishankar** [comment]
I might need to rewrite some of the sprite code to move the ImageData color
switching code to image loading time instead of doing it every time the
sprite is drawn... Its a horrible performance killer.

 I'll end up with hidden canvas's for each sprite with both red and yellow
units on the hidden canvas... Having too many sprites will make it
inefficient.. I think it might make more sense to rewrite the code a little
so we can use one single sprite for each vehicle and one single sprite for
each building. Will be a little bit of work :(

If its OK with you, hold off on doing more work on this and I'll make the
changes using your code changes as a reference and clean up the code at the
same time....

---

**Comment by jakemac53** [comment]
Ok sure sounds good to me. My fork is here https://github.com/jakemac53/command-and-conquer so you can review how I did it. Note that since I only implemented it for one unit type (vehicles) I made a duplicate of the drawSprite function that I just called drawActualSprite. This function accepts the extra parameters required to make the sprite drawing possible (width, height, xoffset).

**Comment by jakemac53** [comment]
Also what may be easier than making hidden canvases (although im not sure if its as efficient) is just keeping an associative array of each color of image in the object instead of just the main sprite, so you might say like this.moveImage['greyscale'] or this.moveImage['original'].

**Comment by adityaravishankar** [comment]
When it comes to multi player, we might need as many as 6 different colors
for each unit (6 teams).... I'll check around to see how efficient hidden
canvas's are.

Drawing from canvas to canvas has the same syntax as image to canvas, so it
won't require big code changes other than storing canvas instead of image
in the original moveImage object and keeping a vertical offset based on the
color we need.

**Comment by jakemac53** [comment]
cool yeah I think its mostly an efficiency question and they are probably fairly similar anyways.

**Comment by adityaravishankar** [comment]
Try playing the game on a slow computer and commenting out the color change
portion of my code.....

 All the units and buildings will be the same color, but the speed at which
the game moves becomes exponentially faster....

Since a lot of people are complaining about performance, I definitely want
to fix this so more people can play the game....

**Comment by jakemac53** [comment]
Oh yeah, as far as efficiency (being fairly similar) I meant saving the other colors on a separate canvas vs as an image. I definitely agree that converting them to greyscale on the fly like it works now is not a good long term solution.

**Comment by jakemac53** [comment]
haha tried playing on my android (1.2ghz single core with 512 ram) and other than the slow framerate and not being able to scroll around its actually somewhat playable rofl.

**Comment by tux-** [comment]
I like to see that you care for image size, and http requests. I tested the mvc.gif image (14512 bytes), converted it to png, and used the optipng on it. Result was 13432 bytes without metadata and 13485 bytes with metadata. Optipng can atleast be found in the debian packages, and very simple to use: optipng -o7 mcv.png

**Comment by jakemac53** [comment]
Ah, never used optipng but I have heard of it before sounds like it does manage to reduce it by close to 10% more than just the standard photoshop options.

**Comment by adityaravishankar** [comment]
Optimized the code.... From 1000+ requests per load to around 100 per load... WOW!!!

 Loading and playing should be much faster now.... 0.3c released at http://www.adityaravishankar.com/projects/games/command-and-conquer/

**Comment by adityaravishankar** [comment]
Wrote a little bit about it on my blog http://www.adityaravishankar.com/2012/01/html5-game-development-using-sprite-sheets-for-better-performance/ :)
